docs/rfcs/optional-config-type.md

```diff
@@ -0,0 +1,184 @@
+}
+```
+
+## Use cases
+
+Optional types can fulfill the following use cases we have when decoding config.
+
+### Representing optional config fields
+
+To use the optional type to mark a config field as optional, the type can simply
+be used as a type parameter to `Optional[T]`. The following config struct shows
+how this may look, both in definition and in usage:
+
+```golang
+type Protocols struct {
+	GRPC Optional[configgrpc.ServerConfig] `mapstructure:"grpc"`
+	HTTP Optional[HTTPConfig]              `mapstructure:"http"`
+}
+
+func (cfg *Config) Validate() error {
+	if !cfg.GRPC.HasValue() && !cfg.HTTP.HasValue() {
+		return errors.New("must specify at least one protocol when using the OTLP receiver")
+	}
+	return nil
+}
+
+func createDefaultConfig() component.Config {
+	return &Config{
+		Protocols: Protocols{
+			GRPC: WithDefault(configgrpc.ServerConfig{
+                // ...
+            }),
+			HTTP: WithDefault(HTTPConfig{
+                // ...
+            }),
+		},
+	}
+}
+```
+
+### Proper unmarshaling of empty values when a default is set
+
+Currently, the OTLP receiver requires a workaround to make enabling each
+protocol optional while providing defaults if a key for the protocol has been
+set:
+
+```golang
+type Protocols struct {
+	GRPC *configgrpc.ServerConfig `mapstructure:"grpc"`
+	HTTP *HTTPConfig              `mapstructure:"http"`
+}
+
+// Config defines configuration for OTLP receiver.
+type Config struct {
+	// Protocols is the configuration for the supported protocols, currently gRPC and HTTP (Proto and JSON).
+	Protocols `mapstructure:"protocols"`
+}
+
+func createDefaultConfig() component.Config {
+	return &Config{
+		Protocols: Protocols{
+			GRPC: configgrpc.NewDefaultServerConfig(),
+			HTTP: &HTTPConfig{
+                // ...
+			},
+		},
+	}
+}
+
+func (cfg *Config) Unmarshal(conf *confmap.Conf) error {
+	// first load the config normally
+	err := conf.Unmarshal(cfg)
+	if err != nil {
+		return err
+	}
+
+    // gRPC will be enabled if this line is not run
+	if !conf.IsSet(protoGRPC) {
+        cfg.GRPC = nil
+	}
+
+     // HTTP will be enabled if this line is not run
+	if !conf.IsSet(protoHTTP) {
+		cfg.HTTP = nil
+	}
+
+	return nil
+}
+```
+
+With an Optional type, the checks in `Unmarshal` become unnecessary, and it's
+possible the entire `Unmarshal` function may no longer be needed. Instead, when
+the the config is unmarshaled, no value would be put into the default Optional
+values and `HasValue` would return false when using this config object.
+
+This situation is something of an edge case with our current unmarshaling
+facilities and while not common, could be a rough edge for users looking to
+implement similar config structures.
+
+## Disadvantages of an Optional type
+
+There are two noteworthy disadvantages of introducing an Optional type:
+
+1. Since the type isn't standard, external packages working with config may
+   require additional adaptations to work with our config structs.
```

Comment:
Can we give examples of external packages working with config? E.g. validation libs come to mind, but they might still work, for them Optional is just an extra nesting container.

Reply:
ok, but how do you want to move the discussion forward? Just take a vote? Or do more cost/benefit analysis, and potentially block this on unresolved #10694?

Reply:
We'll essentially need a vote to get this merged anyway per the Collector's [RFC process](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/rfcs/README.md#merge-rules), so I think that's what it will come down to.

I think this is far more straightforward than https://github.com/open-telemetry/opentelemetry-collector/pull/10694 anyway, so I would prefer to not block it on that. I'm just calling it out here as a risk.

Reply:
I think there is no specific concern on this thread, and the JSON schema libraries support custom types so I am resolving this comment